pkg/controller/jobframework/reconciler.go

```diff
@@ -1148,24 +1150,61 @@ func (r *JobReconciler) ignoreUnretryableError(log logr.Logger, err error) error
-	// from Ready to Completed. As pods finish, they transition first into the
-	// uncountedTerminatedPods staging area, before passing to the
-	// succeeded/failed counters.
-	if workload.IsAdmitted(wl) && (job.PodsReady() || apimeta.IsStatusConditionTrue(wl.Status.Conditions, kueue.WorkloadPodsReady)) {
-		conditionStatus = metav1.ConditionTrue
-		message = "All pods were ready or succeeded since the workload admission"
-	}
-	return metav1.Condition{
-		Type:    kueue.WorkloadPodsReady,
-		Status:  conditionStatus,
-		Reason:  "PodsReady",
-		Message: message,
-		// ObservedGeneration is added via workload.UpdateStatus
+func generatePodsReadyCondition(log logr.Logger, job GenericJob, wl *kueue.Workload) metav1.Condition {
+	if !workload.IsAdmitted(wl) {
+		// The workload has not been admitted yet
+		// or it was admitted in the past but it's evicted/requeued
+		return workload.CreatePodsReadyCondition(metav1.ConditionFalse,
+			kueue.WorkloadWaitForPodsStart,
+			"Not all pods are ready or succeeded")
+	}
+
+	podsReadyCond := apimeta.FindStatusCondition(wl.Status.Conditions, kueue.WorkloadPodsReady)
+	podsReady := job.PodsReady()
+	log.V(3).Info("Generating PodsReady condition",
+		"Current PodsReady condition", podsReadyCond,
+		"Pods are ready", podsReady)
+	switch {
+	case !podsReady && podsReadyCond == nil:
+		return workload.CreatePodsReadyCondition(metav1.ConditionFalse,
+			kueue.WorkloadWaitForPodsStart,
+			"Not all pods are ready or succeeded")
+
+	case podsReady && podsReadyCond == nil:
+		return workload.CreatePodsReadyCondition(metav1.ConditionTrue,
+			kueue.WorkloadWaitForPodsStart,
+			"All pods reached readiness and the workload is running")
+
+	case !podsReady && (podsReadyCond.Reason == kueue.WorkloadWaitForPodsStart || podsReadyCond.Reason == "PodsReady") && podsReadyCond.Status == metav1.ConditionFalse:
+		return workload.CreatePodsReadyCondition(metav1.ConditionFalse,
+			kueue.WorkloadWaitForPodsStart,
+			"Not all pods are ready or succeeded")
+
+	case podsReady && (podsReadyCond.Reason == kueue.WorkloadWaitForPodsStart || podsReadyCond.Reason == "PodsReady"):
+		return workload.CreatePodsReadyCondition(metav1.ConditionTrue,
+			kueue.WorkloadWaitForPodsStart,
+			"All pods reached readiness and the workload is running")
+
+	// a) The workload was running and a pod has failed
+	// b) A pod has failed and the Workload has already recovered once before
+	case !podsReady && ((podsReadyCond.Reason == kueue.WorkloadWaitForPodsStart || podsReadyCond.Reason == "PodsReady") && podsReadyCond.Status == metav1.ConditionTrue ||
+		podsReadyCond.Reason == kueue.WorkloadWaitForPodsRecovery):
+		return workload.CreatePodsReadyCondition(metav1.ConditionFalse,
+			kueue.WorkloadWaitForPodsRecovery,
+			"At least one pod has failed, waiting for recovery")
+
+	case podsReady && podsReadyCond.Reason == kueue.WorkloadWaitForPodsRecovery:
+		return workload.CreatePodsReadyCondition(metav1.ConditionTrue,
+			kueue.WorkloadWaitForPodsRecovery,
```

Comment:
@PBundyra @tenzen-y I find these reasons `WaitForPodsRecovery` and `WaitForPodsStart` suboptimal for the positive scenarios (when pods are ready). Shouldn't we have dedicated "PodsRecovered" and "PodsStarted" reasons?

Reply:
I see, added